src/v/serde/is_serde_compatible.h

```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+#include "reflection/type_traits.h"
+#include "serde/envelope.h"
+#include "serde/serde_is_enum.h"
+#include "tristate.h"
+#include "utils/fragmented_vector.h"
+#include "utils/uuid.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/net/inet_address.hh>
+
+#include <absl/container/btree_map.h>
+#include <absl/container/btree_set.h>
+#include <absl/container/flat_hash_map.h>
+#include <absl/container/flat_hash_set.h>
+#include <absl/container/node_hash_map.h>
+#include <absl/container/node_hash_set.h>
+
+namespace serde {
+
+namespace detail {
+
+template<class T, template<class, size_t> class C>
+struct is_specialization_of_sized : std::false_type {};
+
+template<template<class, size_t> class C, class T, size_t N>
+struct is_specialization_of_sized<C<T, N>, C> : std::true_type {};
+
+template<typename T, template<class, size_t> class C>
+inline constexpr bool is_specialization_of_sized_v
+  = is_specialization_of_sized<T, C>::value;
+
+} // namespace detail
+
+template<typename T>
+concept is_absl_btree_map
+  = ::detail::is_specialization_of_v<T, absl::btree_map>;
+
+template<typename T>
+concept is_absl_flat_hash_map
+  = ::detail::is_specialization_of_v<T, absl::flat_hash_map>;
+
+template<typename T>
+concept is_absl_btree_set
+  = ::detail::is_specialization_of_v<T, absl::btree_set>;
+
+template<typename T>
+concept is_absl_flat_hash_set
+  = ::detail::is_specialization_of_v<T, absl::flat_hash_set>;
+
+template<typename T>
+concept is_fragmented_vector
+  = detail::is_specialization_of_sized_v<T, fragmented_vector>;
+
+template<typename T>
+concept is_chunked_fifo
+  = detail::is_specialization_of_sized_v<T, ss::chunked_fifo>;
+
+template<typename T>
+concept is_std_unordered_map
+  = ::detail::is_specialization_of_v<T, std::unordered_map>;
+
+template<typename T>
+concept is_absl_node_hash_set
+  = ::detail::is_specialization_of_v<T, absl::node_hash_set>;
+
+template<typename T>
+concept is_absl_node_hash_map
+  = ::detail::is_specialization_of_v<T, absl::node_hash_map>;
+
+template<class T>
+concept is_chrono_duration
+  = ::detail::is_specialization_of_v<T, std::chrono::duration>;
+
+template<typename T>
+inline constexpr auto const is_serde_compatible_v
```

Comment:
Is it possible for this to check if `T` has the tag type defined for both read + write instead of this manual list? This file still has the downside of requiring to include everything.

Reply:
`is_serde_compatible.h` is only included in one place. I don't think it helps much. The opposite is true: if we don't check `is_serde_compatible<T>` before calling read or write, we get more detailed error messages about which functions the compiler tried and why they didn't work. So the goal would be to remove this header completely and say "it's serde-compatible if it compiles". There's no need to check it separately.

However, this PR should only do a (more or less) 1:1 transformation from old-serde to tag-invoke-serde. I think it makes more sense to create another PR later that eliminates `is_serde_compatible.h` completely.

Reply:
Update: I added a commit that removes `is_serde_compatible.h` completely. It was only used in one file and is not required as it only prevents the compiler from telling what is the real issue.